imbi/endpoints/project_activity_feed.py

```diff
@@ -1,57 +1,114 @@
+        r'\s+', ' ', """
         WITH created AS (
-                SELECT created_at AS "when",
-                       created_by AS who,
-                       'created' AS what,
-                       NULL AS fact_name,
-                       NULL AS value
-                  FROM v1.projects
-                 WHERE id = %(project_id)s),
-             updated AS (
-                SELECT last_modified_at AS "when",
-                       last_modified_by AS who,
-                       'updated' AS what,
-                       NULL AS fact_name,
-                       NULL AS value
-                  FROM v1.projects
-                 WHERE id = %(project_id)s),
-             facts AS (
-                SELECT a.recorded_at AS "when",
-                       a.recorded_by AS who,
-                       'updated fact' AS what,
-                       b.name AS fact_name,
-                       a.value
-                  FROM v1.project_facts AS a
-                  JOIN v1.project_fact_types AS b
-                    ON b.id = a.fact_type_id
-                 WHERE a.project_id = %(project_id)s),
-            combined AS (
-                SELECT *
-                  FROM created
-                 UNION
-                SELECT *
-                  FROM updated
-                 UNION
-                SELECT *
-                  FROM facts)
-        SELECT a.when,
-               a.who,
-               b.display_name,
-               a.what,
-               a.fact_name,
-               a.value,
-               'ProjectFeedEntry' as "type"
-          FROM combined AS a
-          JOIN v1.users AS b
-            ON b.username = a.who
-         ORDER BY a.when DESC;""")
+          SELECT created_at AS "when", created_by AS who,
+                 'created' AS what, NULL AS fact_name, NULL AS value
+            FROM v1.projects
+           WHERE id = %(project_id)s
+        ),
+        updated AS (
+          SELECT last_modified_at AS "when", last_modified_by AS who,
+                 'updated' AS what, NULL AS fact_name, NULL AS value
+            FROM v1.projects
+           WHERE id = %(project_id)s
+        ),
+        facts AS (
+          SELECT a.recorded_at AS "when", a.recorded_by AS who,
+                 'updated fact' AS what, b.name AS fact_name, a.value
+            FROM v1.project_facts AS a
+            JOIN v1.project_fact_types AS b
+              ON b.id = a.fact_type_id
+           WHERE a.project_id = %(project_id)s
+        ),
+        combined AS (
+                SELECT * FROM created
+          UNION SELECT * FROM updated
+          UNION SELECT * FROM facts
+        )
+        SELECT 'ProjectFeedEntry' AS "type", c."when", c.who, u.display_name,
+               c.what, c.fact_name, c.value
+          FROM combined AS c
+          JOIN v1.users AS u ON u.username = c.who
+         WHERE c."when" BETWEEN %(earlier)s AND %(later)s
+         ORDER BY c."when" DESC
+         LIMIT %(remaining)s
+        """)
```

Comment:
Upon reading this, I noticed we're not actually using the `/projects/<id>/feed` endpoint to grab the project feed 🤦🏽 [I had it grab directly from `/projects/<id>/fact-history`](https://github.com/AWeber-Imbi/imbi-ui/blob/0d9df2149241390b9d6eec1bd6ce5184bccf4224/src/js/views/Project/ProjectFeed/ProjectFeed.jsx#L55).

But this has got me thinking, we can probably remove the `facts` CTE here, and generate another query for fact history directly, and give it a new type `ProjectFactHistory`, and add this query to the `asyncio.gather`. That way we can reuse the existing UI element for the `ProjectFactHistory` elements, while making new ones for the others.

Reply:
And we'll similarly need to grab `v1.users.display_name` in there queries.

Reply:
Let's look at that in another PR.  I agree that we should be retrieving the project specific feed from the fact history table.  There is something subtly different in the queries.  I initially used `v1.activity_history` here filtering on project ID and it failed.  Instead of debugging it further I decided to go with the query that we had.

Reply:
Sounds good.